astropy/stats/funcs.py

```diff
@@ -790,3 +791,133 @@ def mad_std(data):
 
     # NOTE: 1. / scipy.stats.norm.ppf(0.75) = 1.482602218505602
     return median_absolute_deviation(data) * 1.482602218505602
+    
+def akaike_information_criterion(ln_likelihood, num_params, num_samples, use_AICc=True):
+    """
+    Calculate the `Akaike Information Criterion (AIC)
+    <https://en.wikipedia.org/wiki/Akaike_information_criterion>`_.
+
+    The AIC is a penalized likelihood useful for model comparison and preventing 
+    overfitting. It is given by:
+
+    .. math::
+
+        \textrm{AIC} = 2k - 2 \\ln(L)
+    
+    and the AICc, which corrects for finite sample size is
+    
+    .. math::
+
+        \textrm{AICc} = 2k + \\frac{2k(k + 1)}{n - k - 1} - 2 \\ln(L)
+
+    where :math:`n` is the number of data samples, :math:`k` is the number of fitted
+    parameters, and :math:`L` is the likelihood.
+
+    Parameters
+    ----------
+    ln_likelihood : float or array-like
+        natural log likelihood or array of log likelihoods.
+    num_params : int or array-like
+        number of parameters per model
+    num_samples: int
+        number of data points or samples
+    use_AICc: Boolean
+        Use AICc or not. In general, it is better to use the AICc.
+    
+    Returns
+    -------
+    result : float or array-like
+        Returns the AIC or AICc for each set of ln_likelihood, num_params, and num_samples.
+
+    Examples
+    --------
+    Let's say you try fitting a collection of data with multiple models. One model
+    has a ln likelihood of -10, and has 2 fitted parameters. Another model has a 
+    ln likelihood of -5, and has 3 fitted parameters. You have 20 data points. 
+    Does the increase in likelihood justify adding another parameter to the model?
+    The AIC doesn't take into account finite number of data points; the AICc does this
+    and is thus preferable to use (and is set as the default).
+    
+    >>> num_samples = 20
+    >>> num_params = np.array([2,3])
+    >>> ln_likelihood = np.array([-10,-5])
+    >>> aicc = akaike_information_criterion(ln_likelihood, num_params, num_samples, True)
+    >>> print(aicc)
+    [ 24.70588235  17.5       ]
+    
+    Note that when doing model selection using the AIC (or AICc), the lowest AIC is the 
+    preferred model. So, the second model, with more parameters, is the preferred model.
+    The relative probability of the kth model is :math:`\\exp(AIC_{min} - AIC_k)`.
+    
+    Also, the choice of which information criterion to use is up to the user. 
+    For a brief discussion of comparing the AIC to the BIC in an astronomical context,
+    see `Section 3.3 of Kuhn et al. 
+    <http://adsabs.harvard.edu/abs/2014ApJ...787..107K>`_.
+    In general, AIC seems preferred over BIC in most astronomy contexts.
+    """
+    aic = 2*num_params - 2*ln_likelihood
+    if not use_AICc:
+        return aic
+    assert np.all(num_samples - num_params - 1 > 0)
```

Comment:
Might be good to have a more descriptive exception here? (e.g., something other than an assert, or at least an assert with a message)